source/state_representation/src/space/SpatialState.cpp

```diff
@@ -21,11 +23,17 @@ void SpatialState::set_reference_frame(const std::string& reference_frame) {
   this->reference_frame_ = reference_frame;
 }
 
-bool SpatialState::is_compatible(const State& state) const {
-  // FIXME: not sure that the names should be compared here
-  bool compatible = (this->get_name() == state.get_name())
-      && (this->reference_frame_ == dynamic_cast<const SpatialState&>(state).reference_frame_);
-  return compatible;
+bool SpatialState::is_incompatible(const State& state) const {
+  try {
+    auto other = dynamic_cast<const SpatialState&>(state);
+    bool compatible =
+        (this->get_name() == other.reference_frame_) || (this->reference_frame_ == other.get_name())
+            || (this->get_reference_frame() == other.reference_frame_);
+    return !compatible;
```

Comment:
This is a bit hard to read but it checks out. Basically the three valid conditions are:

1. `A` name matches `B` reference_frame (A is parent transform of B)
2. `A` reference frame matches `B` name (A is child transform of B)
3. `A` reference frame matches `B` reference frame (A is sibling transform of B)

You could do it in one line which is maybe even less legible. It might be nice to leave the three cases above as comments just so the reasoning is explicit.

```c++
return (this->get_reference_frame() != other.reference_frame_) && (this->get_name() != other.reference_frame_) && (this->reference_frame_ != other.get_name())
```